pkg/awsutils/awsutils.go

```diff
@@ -347,20 +392,66 @@ func (cache *EC2InstanceMetadataCache) initWithEC2Metadata() error {
 	}
 	log.Debugf("Found vpc-ipv4-cidr-block: %s ", cache.vpcIPv4CIDR)
 
-	// retrieve vpc-ipv4-cidr-blocks
+	// initSGIDs retrieves security groups
+	err = cache.refreshSGIDs(mac)
+	if err != nil {
+		return err
+	}
+
+	// initVPCIPv4CIDRs retrieves VPC IPv4 CIDR blocks
+	err = cache.refreshVPCIPv4CIDRs(mac)
+	if err != nil {
+		return err
+	}
+
+	// refresh security groups and VPC CIDR blocks in the background
+	// Ignoring errors since we will retry in 30s
+	go wait.Forever(func() { _ = cache.refreshSGIDs(mac) }, 30*time.Second)
+	go wait.Forever(func() { _ = cache.refreshVPCIPv4CIDRs(mac) }, 30*time.Second)
+
+	select {
+	case <-ctx.Done():
+		return nil
+	}
+}
+
+// refreshSGIDs retrieves security groups
+func (cache *EC2InstanceMetadataCache) refreshSGIDs(mac string) error {
+	metadataSGIDs, err := cache.ec2Metadata.GetMetadata(metadataMACPath + mac + metadataSGs)
+	if err != nil {
+		awsAPIErrInc("GetMetadata", err)
+		log.Errorf("Failed to retrieve security-group-ids data from instance metadata service")
+		return errors.Wrap(err, "get instance metadata: failed to retrieve security-group-ids")
+	}
+
+	sgIDs := strings.Fields(metadataSGIDs)
+
+	if cache.securityGroups.IsEmpty() {
+		cache.securityGroups.Set(sgIDs)
+	} else {
+		cache.securityGroups.compareAddDel(sgIDs)
+	}
```

Comment:
Just calling `cache.securityGroups.compareAddDel(sgIDs)` directly should also work, or am I missing something?

Reply:
Actually, the above should just be:

```go
newSGs := &StringSet{}.Set(sgIDs...)
deleted := cache.securityGroups.Difference(newSGs)
added := newSGs.Difference(cache.securiyGroups)
if !added.IsEmpty() {
    for _, sg := range added.AWSStrings() {
        log.Infof("Found new %s, added to ipamd cache", *sg)
    }
}
if !deleted.IsEmpty() {
    for _, sg := range deleted.AWSStrings() {
        log.Infof("Removed %s from ipamd cache", *sg)
    }
}
cache.securityGroups.Set(sgIDs)
```